refactor(formatter): improve printing call arguments with arrow function and function expression

Dunqing opened the pull request:
Refactor printing call arguments, which will diverge from the Biome implementation. The main reason I want to do this is that the Biome's implementation is not fit with our printing comments architecture.  

The reason is that our architecture has to ensure that everything is linearly printed, that is, you can't print the second argument and then print the first argument, because printing the second argument might cause the comments of the first argument to attach to the second argument.

Why? Let's take a look at the logic of leading comments.
 
https://github.com/oxc-project/oxc/blob/91bb0f6b8a28afef9b5b7e2352c5face1f0edbab/crates/oxc_formatter/src/formatter/trivia.rs#L100-L106

The implementation regards all comments before this node span as leading comments. I think you may know what will happen when you print the second argument; the comments before the second argument will be fully printed out. then the first argument would miss its comments.

So this refactor aims to solve the above problem. 

Additionally, in this refactor, I am trying to remove unneeded allocations and abstractions as much as possible. Now we have less code and 1% performance improvement. 

<img width="694" height="268" alt="image" src="https://github.com/user-attachments/assets/f533f3e9-8258-4dfd-a4a6-a76b9c425c08" />

Boshen commented:
### Merge activity

* **Jul 21, 9:57 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
refactor(formatter): improve printing call arguments with arrow function and function expression (#12422)

Refactor printing call arguments, which will diverge from the Biome implementation. The main reason I want to do this is that the Biome's implementation is not fit with our printing comments architecture.

The reason is that our architecture has to ensure that everything is linearly printed, that is, you can't print the second argument and then print the first argument, because printing the second argument might cause the comments of the first argument to attach to the second argument.

Why? Let's take a look at the logic of leading comments.

https://github.com/oxc-project/oxc/blob/91bb0f6b8a28afef9b5b7e2352c5face1f0edbab/crates/oxc_formatter/src/formatter/trivia.rs#L100-L106

The implementation regards all comments before this node span as leading comments. I think you may know what will happen when you print the second argument; the comments before the second argument will be fully printed out. then the first argument would miss its comments.

So this refactor aims to solve the above problem.

Additionally, in this refactor, I am trying to remove unneeded allocations and abstractions as much as possible. Now we have less code and 1% performance improvement.

<img width="694" height="268" alt="image" src="https://github.com/user-attachments/assets/f533f3e9-8258-4dfd-a4a6-a76b9c425c08" />
